src/Chess.php

```diff
@@ -521,7 +521,12 @@ protected function undoMove(): ?Move
         $them = self::swapColor($us);
 
         $this->board[$move->from] = $this->board[$move->to];
-        $this->board[$move->from]->type = $move->piece; // to undo any promotions
+        // to undo any promotions
+        $oldFrom = $this->board[$move->from];
+        if (null !== $oldFrom) {
+            $oldFrom->type = $move->piece;
+        }
+        $this->board[$move->from] = $oldFrom;
```

Comment:
what about:
```php
$this->board[$move->from] !== null ? $this->board[$move->from]->type = $move->piece : null;
```

Reply:
The problem here is not possible null value (maybe I can even remove that check), but trying to assign a value to internal array of an ArrayAccess object.